src/nifs/protogalaxy/poly/mod.rs

```diff
@@ -162,7 +173,134 @@ pub(crate) fn compute_F<C: CurveAffine>(
             Ok(UnivariatePoly(points))
         }
         Some(Err(err)) => Err(err.into()),
-        _ => unreachable!(),
+        other => unreachable!("this case must be unreachable: {other:?}"),
+    }
+}
+
+/// This function calculates G(X), which mathematically looks like this:
+///
+/// $$G(X)=\sum_{i=0}^{n-1}\operatorname{pow}_i(\boldsymbol{\beta}+\alpha\cdot\boldsymbol{\delta})f_i(L_0(X)w+\sum_{j\in[k]}L_j(X)w_j)$$
+///
+/// - `f_i` - iteratively all gates for all rows sequentially. The order is taken from
+///           [`plonk::iter_evaluate_witness`].
+/// - `pow_i` - `i` degree of challenge
+/// - `L` - lagrange poly
+///
+/// # Algorithm
+///
+/// We use [`Itertools::tree_reduce`] & store in each node `X` points, for each X challenge
+///
+/// I.e. item `i` from this iterator is a collection of [pow_i(X0), pow_i(X1), ...]
+///
+/// f₀  f₁ f₂  f₃  f₄  f₅  f₆  f₇
+/// │   │  │   │   │   │   │   │
+/// 1   β  1   β   1   β   1   β
+/// │   │  │   │   │   │   │   │
+/// └───f₀₁└───f₂₃ └───f₄₅ └───f₆₇
+///     │      │       │       │
+///     1      β²      1       β²
+///     │      │       │       │
+///     └──────f₀₁₂₃   └───────f₄₅₆₇
+///            │               │
+///            1               β⁴
+///            │               │
+///            └───────────────f₀₁₂₃₄₅₆₇
+///
+/// Each β here is a `beta_stroke` argument
```

Comment:
in this case, the beta_stroke, it is not like: beta_stroke, beta_stroke^2 etc.
it is better use a slightly different version:
`for beta_stroke=(beta_1,beta_2,beta_3)`
`1, beta_1,1, beta_1,1, beta_1,1, beta_1,`
`1,beta_2,1,beta_2`
`1, beta_3`

Reply:
I may have lost the context due to the interruption to benchmark, but could you describe your thought in more detail as it might appear in the doc? So far I haven't understood you

Reply:
Oh, got it from #288 comment, in that case - mistake in this PR

We need something like
```rust 
struct BetaStrokeIter<F> {
    beta: F,
    delta: F,
    X: F,
}
impl<F: PrimeField> Iterator for BetaStrokeIter<F> {
    type Item = F;
    fn next(&mut self) -> Option<Self::Item> {
        let next = self.beta + (self.X * self.delta);
                                                      
        self.beta = self.beta.double();
        self.delta = self.delta.double();
                                                      
        Some(next)
    }
}
```
will fix!

Reply:
Fixed!